staticaddr/loopin/manager.go

```diff
@@ -0,0 +1,336 @@
+	newBlockChan, newBlockErrChan, err := registerBlockNtfn(m.runCtx)
+	if err != nil {
+		return err
+	}
+
+	// Upon start of the loop-in manager we reinstate all previous loop-ins
+	// that are not yet completed.
+	err = m.recover()
+	if err != nil {
+		return err
+	}
+
+	// Communicate to the caller that the address manager has completed its
+	// initialization.
+	close(m.initChan)
+
+	for {
+		select {
+		case height := <-newBlockChan:
+			m.Lock()
+			m.currentHeight = uint32(height)
+			m.Unlock()
+
+		case err = <-newBlockErrChan:
+			return err
+
+		case <-m.runCtx.Done():
+			return m.runCtx.Err()
+		}
+	}
+}
+
+// recover stars a loop-in state machine for each non-final loop-in to pick up
+// work where it was left off before the restart.
+func (m *Manager) recover() error {
+	log.Infof("Recovering static address loop-ins...")
+
+	// Recover loop-ins.
+	loopIns, err := m.cfg.Store.AllLoopIns(m.runCtx)
+	if err != nil {
+		return err
+	}
+
+	for _, loopIn := range loopIns {
+		// If the current deposit is final it wasn't active when we
+		// shut down the client last. So we don't need to start a fsm
+		// for it.
+		if loopIn.IsInFinalState() {
+			continue
+		}
+
+		log.Debugf("Recovering loopIn %x", loopIn.SwapHash)
+
+		// Retrieve all deposits regardless of deposit state.
+		loopIn.Deposits, _ = m.cfg.DepositManager.AllStringOutpointsActiveDeposits( //nolint:lll
+			loopIn.DepositOutpoints, fsm2.EmptyState,
+		)
+
+		loopIn.AddressParams, err = m.cfg.AddressManager.GetStaticAddressParameters( //nolint:lll
+			m.runCtx,
+		)
+		if err != nil {
+			return err
+		}
+
+		loopIn.Address, err = m.cfg.AddressManager.GetStaticAddress(
+			m.runCtx,
+		)
+		if err != nil {
+			return err
+		}
+
+		// Create a state machine for a given loop-in.
+		recovery := true
+		fsm, err := NewFSM(m.runCtx, loopIn, m.cfg, recovery)
+		if err != nil {
+			return err
+		}
+
+		// Send the OnRecover event to the state machine.
+		go func() {
+			err = fsm.SendEvent(OnRecover, nil)
+			if err != nil {
+				log.Errorf("Error sending OnStart event: %v",
+					err)
+			}
+		}()
+	}
+
+	return nil
+}
+
+// WaitInitComplete waits until the static address loop-in manager has completed
+// its setup.
+func (m *Manager) WaitInitComplete() {
+	defer log.Debugf("Static address loop-in manager initiation complete.")
+	<-m.initChan
+}
+
+// InitiateLoopIn initiates a loop-in swap. It passes the request to the server
+// along with all relevant loop-in information.
+func (m *Manager) InitiateLoopIn(req *loop.StaticAddressLoopInRequest) error {
+	loopIn := &StaticAddressLoopIn{}
+
+	// Validate the loop-in request.
+	if len(req.DepositOutpoints) == 0 {
+		return fmt.Errorf("no deposit outpoints provided")
+	}
+	loopIn.DepositOutpoints = req.DepositOutpoints
+
+	// Retrieve all deposits referenced by the outpoints and ensure that
+	// they are in state Deposited.
+	deposits, active := m.cfg.DepositManager.AllStringOutpointsActiveDeposits( //nolint:lll
```

Comment:
Since this function is reentrant, how can we ensure that the deposits are not being used by another concurrent loopin fsm?

Reply:
lgtm